### PR TITLE
Custom display API

### PR DESCRIPTION
**Pull request by jerrylian-db**
* Motivation for features / changes
    - This feature lets users to directly set an inline display function for the notebook API that works in their environment, providing more flexibility for Tensorboard to be used on various notebook environments. See RFC https://github.com/tensorflow/tensorboard/issues/3752 for more details.

* Technical description of changes
    - Added a public API to the notebook module that allows users to set a custom display function.
    -  In `_get_context` ([link](https://github.com/jerrylian-db/tensorboard/blob/ed147694346da9c1553c2da1a705cc03c38b3298/tensorboard/notebook.py#L60)), we add a context type that is returned when the library detects that the custom display function is set. This new context type then causes the `_display` function ([link](https://github.com/jerrylian-db/tensorboard/blob/ed147694346da9c1553c2da1a705cc03c38b3298/tensorboard/notebook.py#L283)) to use the custom display function to display Tensorboard inline to python notebooks.

* Screenshots of UI changes
    - N/A

* Detailed steps to verify changes work correctly (as executed by you)
    1. Clone the repo.
    2. In the repo root directory, start a Jupyter notebook process and create a notebook inside of it. (Make sure that your environment has Tensorboard installed as the following code depends on having a Tensorboard binary in the environment.)
    3. Run the following code..
```python
import tensorboard
import random, IPython
def custom_display(port, height, display_handle):
    del display_handle

    frame_id = "tensorboard-frame-{:08x}".format(random.getrandbits(64))
    html_string = """
        <h1>Custom Inline Display</h1>
        <iframe id="%ID%" width="100%" height="%HEIGHT%" frameborder="0"></iframe>
        <script>
            (function() {
                const frame = document.getElementById("%ID%");
                const url = new URL("/", window.location);
                const port = %PORT%;
                if (port) {
                  url.port = port;
                }
                frame.src = url;
            })();
        </script>
    </html>
    """

    replacements = [
        ("ID%", frame_id),
        ("%HEIGHT%", "%d" % height),
        ("%PORT%", "%d" % port),
    ]

    for (k, v) in replacements:
        html_string = html_string.replace(k, v)
    iframe = IPython.display.HTML(html_string)
    IPython.display.display(iframe)
tensorboard.notebook.register_custom_display(custom_display)
tensorboard.notebook.start("--logdir test")
```
Result:
Screen Shot 2020-06-29 at 3 49 40 PM

* Alternate designs / implementations considered
    - We considered directly modifying the `_display` function to return the custom display if it detects it is set, rather than modifying `_get_context`. But we thought that the presented solution matches existing patterns better.

### CODE REVIEW

**Comment by jerrylian-db** [comment]
Note: I did not find any automated tests for the `notebook.py` file. Please tell me if and how I should add test coverage for this feature.

**Comment by jerrylian-db** [comment]
@googlebot I signed it!

File: tensorboard/notebook.py [hunk 1]
```diff
@@ -145,7 +158,7 @@ def start(args_string):
     except ImportError:
         IPython = None
 
-    if context == _CONTEXT_NONE:
+    if context == _CONTEXT_NONE or context == _CONTEXT_CUSTOM:
```
Comment (falaki) [inline tensorboard/notebook.py:161]: Why should we print `Launching TensorBoard...` when custom display function has been registered?

File: tensorboard/notebook.py [hunk 2]
```diff
@@ -49,19 +49,32 @@
 _CONTEXT_COLAB = "_CONTEXT_COLAB"
 _CONTEXT_IPYTHON = "_CONTEXT_IPYTHON"
 _CONTEXT_NONE = "_CONTEXT_NONE"
+_CONTEXT_CUSTOM = "_CONTEXT_CUSTOM"
+
+_custom_display = None
+
+
+def register_custom_display(func):
```
Comment (falaki) [inline tensorboard/notebook.py:57]: Nit: add some documentation for this function.

Comment (falaki) [reply tensorboard/notebook.py:57]: Specifically mention what is the signature of `func` argument.

File: tensorboard/notebook.py (hunk 1 above)
Comment (jerrylian-db) [reply tensorboard/notebook.py:161]: This tells the user that the Tensorboard process is launching. It is standard behavior across the different python environments.

**Comment by jerrylian-db** [comment]
@wchargin can you take a look at this? Thanks for your time.

**Comment by wchargin** [comment]
@jerrylian-db yep—responded on the RFC:
<https://github.com/tensorflow/tensorboard/issues/3752#issuecomment-662622316>

**Pull request closed without merging**